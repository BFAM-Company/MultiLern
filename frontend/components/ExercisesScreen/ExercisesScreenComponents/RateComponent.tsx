import React from 'react'
import { Image, ImageBackground, StyleSheet, Text, TouchableOpacity, View } from 'react-native'
import YellowStar from './YellowStar';


interface RateComponentProps {
  rate: number
}

function RateComponent({rate}: RateComponentProps) {

  let rates = [0,0,0,0,0]
  for(let i = 0; i<rates.length; i++){
    if(rate >= (i+1)){
        rates[i] = 1
    }
    else{
        rates[i] = rate%1;
        break;
    }
  }

  return (
    <View
        style={styles.container}
    >
        <YellowStar rate={rates[0]}/>
        <YellowStar rate={rates[1]}/>
        <YellowStar rate={rates[2]}/>
        <YellowStar rate={rates[3]}/>
        <YellowStar rate={rates[4]}/>
    </View>
  );
}




const styles = StyleSheet.create({
    container:{
        width:'100%',
        height:'100%',
        display:'flex',
        flexDirection:'row',
        alignItems:'center',
        justifyContent:'space-around'
    },
})

export default RateComponent;