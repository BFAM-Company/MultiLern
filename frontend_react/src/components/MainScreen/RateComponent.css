.rate-container {
  display: flex;
  flex-direction: row;
  justify-content: space-around;
  align-items: center;
  width: 100%;
  height: 100%;
}
