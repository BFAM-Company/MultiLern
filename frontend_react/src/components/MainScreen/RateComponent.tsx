import React, { useEffect, useState } from 'react';
import './RateComponent.css';
import YellowStar from './YellowStar.tsx';

interface RateComponentProps {
  rate: number;
}

function RateComponent({ rate }: RateComponentProps) {
  const [currentRates, setCurrentRates] = useState<number[]>([0, 0, 0, 0, 0]);

  const calcStars = (rating: number) => {
    let rates = [0, 0, 0, 0, 0];

    for (let i = 0; i < rates.length; i++) {
      if (rating >= i + 1) {
        rates[i] = 1;
      } else {
        rates[i] = rating % 1;
        break;
      }
    }
    setCurrentRates(rates);
  };

  const userRateEvent = async (value: number) => {
    // Perform any fetch or logic here
    calcStars(value);
  };

  useEffect(() => {
    calcStars(rate);
  }, [rate]);

  return (
    <div className="rate-container">
      <YellowStar rate={currentRates[0]} handlePress={() => userRateEvent(1)} />
      <YellowStar rate={currentRates[1]} handlePress={() => userRateEvent(2)} />
      <YellowStar rate={currentRates[2]} handlePress={() => userRateEvent(3)} />
      <YellowStar rate={currentRates[3]} handlePress={() => userRateEvent(4)} />
      <YellowStar rate={currentRates[4]} handlePress={() => userRateEvent(5)} />
    </div>
  );
}

export default RateComponent;
