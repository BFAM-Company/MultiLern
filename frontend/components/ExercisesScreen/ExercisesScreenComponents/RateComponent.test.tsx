import { fireEvent, render, waitFor } from "@testing-library/react-native";
import React from "react";
import RateComponent from "./RateComponent";

jest.mock("@react-native-async-storage/async-storage", () => require("../../../__mocks__/mock-async-storage"));


describe('RateComponent', () => {
    afterEach(() => {
        jest.clearAllMocks();
    });

    it('renders correctly', () => {
        const { toJSON } = render(<RateComponent rate={5} />);
        expect(toJSON()).toMatchSnapshot();
    });


    it('displays correct number of stars based on initial rate prop', async () => {
        const { getAllByTestId } = render(<RateComponent rate={3} />);
        const stars = getAllByTestId('star')

        expect(stars[0]).toHaveStyle({ opacity: 1 });
        expect(stars[1]).toHaveStyle({ opacity: 1 });
        expect(stars[2]).toHaveStyle({ opacity: 1 });
        expect(stars[3]).toHaveStyle({ opacity: 1 });
        expect(stars[4]).toHaveStyle({ opacity: 1 });
    });

    it('updates stars when user rates', async () => {
        const { getAllByTestId } = render(<RateComponent rate={3} />);
        const star5 = getAllByTestId('star')[4];

        fireEvent.press(star5);

        await waitFor(() => {
            const updatedStar5 = getAllByTestId('star')[4];
            expect(updatedStar5).toHaveStyle({ opacity: 1 });
        });
    });

});
