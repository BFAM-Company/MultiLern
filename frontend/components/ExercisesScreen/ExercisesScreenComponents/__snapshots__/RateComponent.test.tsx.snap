// Jest Snapshot v1, https://goo.gl/fbAQLP

exports[`RateComponent renders correctly 1`] = `
<View
  style={
    {
      "alignItems": "center",
      "display": "flex",
      "flexDirection": "row",
      "height": "100%",
      "justifyContent": "space-around",
      "width": "100%",
    }
  }
>
  <View
    accessibilityState={
      {
        "busy": undefined,
        "checked": undefined,
        "disabled": undefined,
        "expanded": undefined,
        "selected": undefined,
      }
    }
    accessibilityValue={
      {
        "max": undefined,
        "min": undefined,
        "now": undefined,
        "text": undefined,
      }
    }
    accessible={true}
    collapsable={false}
    focusable={true}
    onClick={[Function]}
    onResponderGrant={[Function]}
    onResponderMove={[Function]}
    onResponderRelease={[Function]}
    onResponderTerminate={[Function]}
    onResponderTerminationRequest={[Function]}
    onStartShouldSetResponder={[Function]}
    style={
      {
        "alignItems": "center",
        "display": "flex",
        "height": 25,
        "opacity": 1,
        "width": 25,
      }
    }
    testID="star"
  >
    <Image
      source={1}
      style={
        [
          {
            "height": 25,
            "left": 0,
            "position": "absolute",
            "resizeMode": "cover",
            "top": 0,
            "width": 25,
          },
          {
            "tintColor": "dimgray",
          },
        ]
      }
    />
    <View
      style={
        [
          {
            "alignItems": "flex-start",
            "display": "flex",
            "justifyContent": "center",
            "overflow": "hidden",
          },
          {
            "bottom": 0,
            "left": 0,
            "position": "absolute",
            "right": 0,
            "top": 0,
            "width": 25,
          },
        ]
      }
    >
      <Image
        source={1}
        style={
          [
            {
              "height": 25,
              "left": 0,
              "position": "absolute",
              "resizeMode": "cover",
              "top": 0,
              "width": 25,
            },
            {
              "tintColor": "gold",
            },
          ]
        }
      />
    </View>
  </View>
  <View
    accessibilityState={
      {
        "busy": undefined,
        "checked": undefined,
        "disabled": undefined,
        "expanded": undefined,
        "selected": undefined,
      }
    }
    accessibilityValue={
      {
        "max": undefined,
        "min": undefined,
        "now": undefined,
        "text": undefined,
      }
    }
    accessible={true}
    collapsable={false}
    focusable={true}
    onClick={[Function]}
    onResponderGrant={[Function]}
    onResponderMove={[Function]}
    onResponderRelease={[Function]}
    onResponderTerminate={[Function]}
    onResponderTerminationRequest={[Function]}
    onStartShouldSetResponder={[Function]}
    style={
      {
        "alignItems": "center",
        "display": "flex",
        "height": 25,
        "opacity": 1,
        "width": 25,
      }
    }
    testID="star"
  >
    <Image
      source={1}
      style={
        [
          {
            "height": 25,
            "left": 0,
            "position": "absolute",
            "resizeMode": "cover",
            "top": 0,
            "width": 25,
          },
          {
            "tintColor": "dimgray",
          },
        ]
      }
    />
    <View
      style={
        [
          {
            "alignItems": "flex-start",
            "display": "flex",
            "justifyContent": "center",
            "overflow": "hidden",
          },
          {
            "bottom": 0,
            "left": 0,
            "position": "absolute",
            "right": 0,
            "top": 0,
            "width": 25,
          },
        ]
      }
    >
      <Image
        source={1}
        style={
          [
            {
              "height": 25,
              "left": 0,
              "position": "absolute",
              "resizeMode": "cover",
              "top": 0,
              "width": 25,
            },
            {
              "tintColor": "gold",
            },
          ]
        }
      />
    </View>
  </View>
  <View
    accessibilityState={
      {
        "busy": undefined,
        "checked": undefined,
        "disabled": undefined,
        "expanded": undefined,
        "selected": undefined,
      }
    }
    accessibilityValue={
      {
        "max": undefined,
        "min": undefined,
        "now": undefined,
        "text": undefined,
      }
    }
    accessible={true}
    collapsable={false}
    focusable={true}
    onClick={[Function]}
    onResponderGrant={[Function]}
    onResponderMove={[Function]}
    onResponderRelease={[Function]}
    onResponderTerminate={[Function]}
    onResponderTerminationRequest={[Function]}
    onStartShouldSetResponder={[Function]}
    style={
      {
        "alignItems": "center",
        "display": "flex",
        "height": 25,
        "opacity": 1,
        "width": 25,
      }
    }
    testID="star"
  >
    <Image
      source={1}
      style={
        [
          {
            "height": 25,
            "left": 0,
            "position": "absolute",
            "resizeMode": "cover",
            "top": 0,
            "width": 25,
          },
          {
            "tintColor": "dimgray",
          },
        ]
      }
    />
    <View
      style={
        [
          {
            "alignItems": "flex-start",
            "display": "flex",
            "justifyContent": "center",
            "overflow": "hidden",
          },
          {
            "bottom": 0,
            "left": 0,
            "position": "absolute",
            "right": 0,
            "top": 0,
            "width": 25,
          },
        ]
      }
    >
      <Image
        source={1}
        style={
          [
            {
              "height": 25,
              "left": 0,
              "position": "absolute",
              "resizeMode": "cover",
              "top": 0,
              "width": 25,
            },
            {
              "tintColor": "gold",
            },
          ]
        }
      />
    </View>
  </View>
  <View
    accessibilityState={
      {
        "busy": undefined,
        "checked": undefined,
        "disabled": undefined,
        "expanded": undefined,
        "selected": undefined,
      }
    }
    accessibilityValue={
      {
        "max": undefined,
        "min": undefined,
        "now": undefined,
        "text": undefined,
      }
    }
    accessible={true}
    collapsable={false}
    focusable={true}
    onClick={[Function]}
    onResponderGrant={[Function]}
    onResponderMove={[Function]}
    onResponderRelease={[Function]}
    onResponderTerminate={[Function]}
    onResponderTerminationRequest={[Function]}
    onStartShouldSetResponder={[Function]}
    style={
      {
        "alignItems": "center",
        "display": "flex",
        "height": 25,
        "opacity": 1,
        "width": 25,
      }
    }
    testID="star"
  >
    <Image
      source={1}
      style={
        [
          {
            "height": 25,
            "left": 0,
            "position": "absolute",
            "resizeMode": "cover",
            "top": 0,
            "width": 25,
          },
          {
            "tintColor": "dimgray",
          },
        ]
      }
    />
    <View
      style={
        [
          {
            "alignItems": "flex-start",
            "display": "flex",
            "justifyContent": "center",
            "overflow": "hidden",
          },
          {
            "bottom": 0,
            "left": 0,
            "position": "absolute",
            "right": 0,
            "top": 0,
            "width": 25,
          },
        ]
      }
    >
      <Image
        source={1}
        style={
          [
            {
              "height": 25,
              "left": 0,
              "position": "absolute",
              "resizeMode": "cover",
              "top": 0,
              "width": 25,
            },
            {
              "tintColor": "gold",
            },
          ]
        }
      />
    </View>
  </View>
  <View
    accessibilityState={
      {
        "busy": undefined,
        "checked": undefined,
        "disabled": undefined,
        "expanded": undefined,
        "selected": undefined,
      }
    }
    accessibilityValue={
      {
        "max": undefined,
        "min": undefined,
        "now": undefined,
        "text": undefined,
      }
    }
    accessible={true}
    collapsable={false}
    focusable={true}
    onClick={[Function]}
    onResponderGrant={[Function]}
    onResponderMove={[Function]}
    onResponderRelease={[Function]}
    onResponderTerminate={[Function]}
    onResponderTerminationRequest={[Function]}
    onStartShouldSetResponder={[Function]}
    style={
      {
        "alignItems": "center",
        "display": "flex",
        "height": 25,
        "opacity": 1,
        "width": 25,
      }
    }
    testID="star"
  >
    <Image
      source={1}
      style={
        [
          {
            "height": 25,
            "left": 0,
            "position": "absolute",
            "resizeMode": "cover",
            "top": 0,
            "width": 25,
          },
          {
            "tintColor": "dimgray",
          },
        ]
      }
    />
    <View
      style={
        [
          {
            "alignItems": "flex-start",
            "display": "flex",
            "justifyContent": "center",
            "overflow": "hidden",
          },
          {
            "bottom": 0,
            "left": 0,
            "position": "absolute",
            "right": 0,
            "top": 0,
            "width": 25,
          },
        ]
      }
    >
      <Image
        source={1}
        style={
          [
            {
              "height": 25,
              "left": 0,
              "position": "absolute",
              "resizeMode": "cover",
              "top": 0,
              "width": 25,
            },
            {
              "tintColor": "gold",
            },
          ]
        }
      />
    </View>
  </View>
</View>
`;
